packages/fresh/src/middlewares/ip_restriction.ts

```diff
@@ -0,0 +1,122 @@
+import type { Context } from "../context.ts";
+import type { Middleware } from "./mod.ts";
+import { isIPv4, isIPv6, matchSubnets } from "@std/net/unstable-ip";
+
+/**
+ * Configuration rules for IP restriction middleware.
+ */
+export interface IPRestrictionRules {
+  /**
+   * List of IP addresses or CIDR blocks to deny access.
+   * If an IP matches any entry in this list, access will be blocked.
+   *
+   * @example ["192.168.1.10", "10.0.0.0/8", "2001:db8::1"]
+   */
+  denyList?: string[];
+
+  /**
+   * List of IP addresses or CIDR blocks to allow access.
+   * If specified, only IPs matching entries in this list will be allowed.
+   * If empty or undefined, all IPs are allowed (unless in denyList).
+   *
+   * @example ["192.168.1.0/24", "203.0.113.0/24", "2001:db8::/32"]
+   */
+  allowList?: string[];
+}
+
+function blockError(): Response {
+  return new Response("Forbidden", { status: 403 });
+}
+
+function findType(
+  addr: string,
+): Deno.NetworkInterfaceInfo["family"] | undefined {
+  if (isIPv4(addr)) {
+    return "IPv4";
+  } else if (isIPv6(addr)) {
+    return "IPv6";
+  }
+  return undefined;
+}
+
+export interface IpRestrictionOptions {
+  onError?: <State>(
+    remote: {
+      addr: string;
+      type: Deno.NetworkInterfaceInfo["family"] | undefined;
+    },
+    ctx: Context<State>,
+  ) => Response | Promise<Response>;
+}
+
+/**
+ * IP restriction Middleware for Fresh.
+ *
+ * @param rules rules `{ denyList: string[], allowList: string[] }`.
+ * @param options Options for the IP Restriction middleware.
+ * @returns The middleware handler function.
+ *
+ * @example Basic usage (with defaults)
+ * ```ts
+ * const app = new App<State>()
+ *
+ * app.use(ipRestriction({denyList: ["192.168.1.10", "2001:db8::1"]}))
+ * ```
+ *
+ * @example Custom error handling
+ * ```ts
+ * const customOnError: IpRestrictionOptions = {
+ *   onError: (remote, ctx) => {
+ *     console.log(
+ *       `Request URL: ${ctx.url}, Blocked IP: ${remote.addr} of type ${remote.type}`,
+ *     );
+ *
+ *     return new Response("custom onError", { status: 401 });
+ *   },
+ * };
+ * app.use(ipRestriction({denyList: ["192.168.1.10", "2001:db8::1"]}, customOnError))
+ * ```
+ */
+export function ipRestriction<State>(
+  rules: IPRestrictionRules,
+  options?: IpRestrictionOptions,
+): Middleware<State> {
+  return async function ipRestriction<State>(ctx: Context<State>) {
+    if (
+      ctx.info.remoteAddr.transport !== "udp" &&
+      ctx.info.remoteAddr.transport !== "tcp"
+    ) {
+      throw new TypeError(
+        "Unsupported transport protocol. TCP & UDP is supported.",
+      );
+    }
```

Comment:
A good question to ask is "does Fresh allow other traffic besides UDP and TCP?" If the answer is no, this section can be removed, as the conditional will never evaluate to `true`.

Reply:
To clarify, I'm wondering whether Fresh does any preprocessing of traffic that prevents non-UDP and non-TCP from being served... But according to #2392, UNIX sockets are supported (I assume it hasn't changed).

Reply:
Actually, I'm thinking it's worth removing this block anyway. I think we should filter IPs regardless of whether it's UDP or TCP traffic or not.

Reply:
Sorry, to correct my stance, yes, we should return early if not UDP or TCP traffic, because an IP address isn't being used. But I don't think throwing is the correct behaviour. Instead, we should just pass through. See my suggestion above.